src/ATen/native/xpu/sycl/UpSampleBicubic2dKernels.cpp

```diff
@@ -0,0 +1,179 @@
+              static_cast<scalar_t>(cubic_interp1d<scalar_t, accscalar_t>(
+                  coefficients[0],
+                  coefficients[1],
+                  coefficients[2],
+                  coefficients[3],
+                  t_y));
+        }
+      }
+    }
+  }
+  UpsampleBicubic2dKernelFunctor(
+      PackedTensorAccessor64<scalar_t, 4> out_data,
+      const PackedTensorAccessor64<const scalar_t, 4> in_data,
+      int64_t onum,
+      bool align_corners,
+      const accscalar_t height_scale,
+      const accscalar_t width_scale)
+      : out_data_(out_data),
+        in_data_(in_data),
+        onum_(onum),
+        align_corners_(align_corners),
+        height_scale_(height_scale),
+        width_scale_(width_scale) {}
+
+ private:
+  PackedTensorAccessor64<scalar_t, 4> out_data_;
+  const PackedTensorAccessor64<const scalar_t, 4> in_data_;
+  int64_t onum_;
+  bool align_corners_;
+  const accscalar_t height_scale_;
+  const accscalar_t width_scale_;
+};
+
+template <typename scalar_t, typename accscalar_t>
+static void upsample_bicubic2d_out_frame(
+    PackedTensorAccessor64<scalar_t, 4> odata,
+    const PackedTensorAccessor64<const scalar_t, 4> idata,
+    int64_t onum,
+    bool align_corners,
+    const accscalar_t height_scale,
+    const accscalar_t width_scale) {
+  auto queue = getCurrentSYCLQueue();
+  int64_t wg_size = syclMaxWorkGroupSize();
+  int64_t wg_num = at::ceil_div(onum, wg_size);
+
+  UpsampleBicubic2dKernelFunctor<scalar_t, accscalar_t> kfn(
+      odata, idata, onum, align_corners, height_scale, width_scale);
+  sycl_kernel_submit(wg_num * wg_size, wg_size, queue, kfn);
+}
+
+void upsample_bicubic2d_kernel(
+    Tensor& output,
+    const Tensor& input,
+    IntArrayRef output_size,
+    bool align_corners,
+    std::optional<double> scales_h,
+    std::optional<double> scales_w) {
+  TensorArg input_arg{input, "input", 1}, output_arg{output, "output", 2};
+  checkAllSameGPU(__func__, {input_arg, output_arg});
+
+  int output_height = output_size[0];
+  int output_width = output_size[1];
+
+  int input_height = input.size(2);
+  int input_width = input.size(3);
+
+  output.zero_();
+
+  const int num_output_elements = output_height * output_width;
+
+  AT_DISPATCH_FLOATING_TYPES_AND2(
+      at::ScalarType::Half,
+      at::ScalarType::BFloat16,
+      input.scalar_type(),
+      "upsample_bicubic2d_kernel",
+      [&] {
+        auto idata = input.packed_accessor64<const scalar_t, 4>();
+        auto odata = output.packed_accessor64<scalar_t, 4>();
+
+        // Get scaling factors
+        using accscalar_t = at::acc_type<scalar_t, true>;
+        const accscalar_t rheight = area_pixel_compute_scale<accscalar_t>(
+            input_height, output_height, align_corners, scales_h);
+        const accscalar_t rwidth = area_pixel_compute_scale<accscalar_t>(
+            input_width, output_width, align_corners, scales_w);
+
+        upsample_bicubic2d_out_frame<scalar_t, accscalar_t>(
```

Comment:
CUDA upsample_bicubic2d_out_frame means a CUDA kernel, not a launcher. Maybe we can name it upsample_bicubic2d_out_template or launch_upsample_bicubic2d_kernel.